firebase-firestore/src/main/java/com/google/firebase/firestore/core/TransactionRunner.java

```diff
@@ -0,0 +1,116 @@
+  private RemoteStore remoteStore;
+  private Function<Transaction, Task<TResult>> updateFunction;
+  private int retriesLeft;
+
+  private ExponentialBackoff backoff;
+  private TaskCompletionSource<TResult> taskSource = new TaskCompletionSource<>();
+
+  public TransactionRunner(
+      AsyncQueue asyncQueue,
+      RemoteStore remoteStore,
+      Function<Transaction, Task<TResult>> updateFunction) {
+
+    this.asyncQueue = asyncQueue;
+    this.remoteStore = remoteStore;
+    this.updateFunction = updateFunction;
+    this.retriesLeft = RETRY_COUNT;
+
+    backoff = new ExponentialBackoff(asyncQueue, TimerId.RETRY_TRANSACTION);
+  }
+
+  /** Runs the transaction and sets the result in taskSource. */
+  public Task<TResult> run() {
+    runWithBackoff();
+    return taskSource.getTask();
+  }
+
+  private void runWithBackoff() {
+    backoff.backoffAndRun(
+        () -> {
+          final Transaction transaction = remoteStore.createTransaction();
+          updateFunction
+              .apply(transaction)
+              .addOnCompleteListener(
+                  asyncQueue.getExecutor(),
+                  new OnCompleteListener<TResult>() {
+                    @Override
+                    public void onComplete(@NonNull Task<TResult> userTask) {
+                      if (!userTask.isSuccessful()) {
+                        handleTransactionError(userTask);
+                      } else {
+                        transaction
+                            .commit()
+                            .addOnCompleteListener(
+                                asyncQueue.getExecutor(),
+                                new OnCompleteListener<Void>() {
+                                  @Override
+                                  public void onComplete(@NonNull Task<Void> commitTask) {
+                                    if (commitTask.isSuccessful()) {
+                                      taskSource.setResult(userTask.getResult());
+                                    } else {
+                                      handleTransactionError(commitTask);
+                                    }
+                                  }
+                                });
+                      }
+                    }
+                  });
+        });
+  }
+
+  private static boolean isRetryableTransactionError(Exception e) {
+    if (e instanceof FirebaseFirestoreException) {
+      // In transactions, the backend will fail outdated reads with FAILED_PRECONDITION and
+      // non-matching document versions with ABORTED. These errors should be retried.
+      FirebaseFirestoreException.Code code = ((FirebaseFirestoreException) e).getCode();
+      return code == FirebaseFirestoreException.Code.ABORTED
+          || code == FirebaseFirestoreException.Code.FAILED_PRECONDITION
+          || !Datastore.isPermanentError(((FirebaseFirestoreException) e).getCode());
+    }
+    return false;
+  }
+
+  private void handleTransactionError(Task task) {
```

Comment:
nit: We typically organize methods "top-down" meaning that this should go above isRetryableTransactionError(), since runWithBackoff() calls handleTransactionError() which calls isRetryableTransactionError().